### PR TITLE
Fixes mozilla/thimble.webmaker.org#893

### PR DESCRIPTION
**Pull request by sedge**

### CODE REVIEW

File: api/classes/base_controller.js [hunk 1]
```diff
@@ -65,11 +65,12 @@ BaseController.prototype.create = function(req, reply) {
 };
 
 BaseController.prototype.delete = function(req, reply) {
-  var result = Promise.resolve().then(function() {
-    var record = req.pre.records.models[0];
-
-    record.destroy();
+  var record = req.pre.records.models[0];
 
+  var result = Promise.resolve().then(function() {
```
Comment (gideonthomas) [inline api/classes/base_controller.js]: I knew it!!!

File: api/classes/publisher.js [hunk 2]
```diff
@@ -323,26 +323,37 @@ exports.publish = function publish(project) {
...
+  }
+
+  function destroyPublishedFiles(user, publishedProject) {
+    function deleteFilesFromS3(files) {
+      return modifyPublishedFiles(deleteFileRemotely, project, user, files);
+    }
+
+    function unsetProjectReferences() {
+      return project.set({
+        published_id: null,
```
Comment (gideonthomas) [inline api/classes/publisher.js]: you should make this an on delete cascade fk

File: api/classes/base_controller.js (hunk 1 above)
Comment (sedge) [inline api/classes/base_controller.js]: I wish I could say this was the problem. It's not the problem haha

File: api/modules/projects/controller.js [hunk 3]
```diff
@@ -57,13 +60,53 @@ controller.delete = function(req, reply) {
...
+  function clearAllPublishedFiles() {
+    function fetchFiles() {
+      return Files.query({
+        where: {
+          project_id: project.get('id')
+        }
+      }).fetchAll();
+    }
+
+    function fetchPublishedFilesByFile(file) {
```
Comment (gideonthomas) [inline api/modules/projects/controller.js]: fetchPublishedFilesByFileId

Comment (gideonthomas) [inline api/modules/projects/controller.js]: I think fetchPublishedFiles is fine

File: api/modules/projects/controller.js [hunk 4]
```diff
@@ -57,13 +60,53 @@ controller.delete = function(req, reply) {
...
+        }
+      }).fetchAll();
+    }
+
+    function clearPublishedFiles(publishedFiles) {
+      if (publishedFiles.length === 0) {
+        return;
+      }
+      return publishedFiles.mapThen(function(file) {
+        return file.destroy();
```
Comment (gideonthomas) [inline api/modules/projects/controller.js]: publishedFile.destroy

**Comment by gideonthomas** [comment]
r+ with nits

**Pull request merged**